homeassistant/components/crownstone/listeners.py

```diff
@@ -0,0 +1,147 @@
+
+
+@callback
+def async_update_crwn_state_sse(
+    manager: CrownstoneEntryManager, ha_event: Event
+) -> None:
+    """Update the state of a Crownstone when switched externally."""
+    switch_event = SwitchStateUpdateEvent(ha_event.data)
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_id(switch_event.cloud_id)
+    except KeyError:
+        return
+
+    # only update on change.
+    if updated_crownstone.state != switch_event.switch_state:
+        updated_crownstone.state = switch_event.switch_state
+        async_dispatcher_send(manager.hass, SIG_CROWNSTONE_STATE_UPDATE)
+
+
+@callback
+def async_update_crwn_ability(manager: CrownstoneEntryManager, ha_event: Event) -> None:
+    """Update the ability information of a Crownstone."""
+    ability_event = AbilityChangeEvent(ha_event.data)
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_id(ability_event.cloud_id)
+    except KeyError:
+        return
+
+    ability_type = ability_event.ability_type
+    ability_enabled = ability_event.ability_enabled
+    # only update on a change in state
+    if updated_crownstone.abilities[ability_type].is_enabled == ability_enabled:
+        return
+
+    # write the change to the crownstone entity.
+    updated_crownstone.abilities[ability_type].is_enabled = ability_enabled
+
+    if ability_event.sub_type == EVENT_ABILITY_CHANGE_DIMMING:
+        # reload the config entry because dimming is part of supported features
+        manager.hass.async_create_task(
+            manager.hass.config_entries.async_reload(manager.config_entry.entry_id)
+        )
+    else:
+        async_dispatcher_send(manager.hass, SIG_CROWNSTONE_STATE_UPDATE)
+
+
+def update_uart_state(manager: CrownstoneEntryManager, _: bool | None) -> None:
+    """Update the uart ready state for entities that use USB."""
+    # update availability of power usage entities.
+    dispatcher_send(manager.hass, SIG_UART_STATE_CHANGE)
+
+
+def update_crwn_state_uart(
+    manager: CrownstoneEntryManager, data: AdvExternalCrownstoneState
+) -> None:
+    """Update the state of a Crownstone when switched externally."""
+    if data.type != AdvType.EXTERNAL_STATE:
+        return
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_uid(
+            data.crownstoneId, manager.usb_sphere_id
+        )
+    except KeyError:
+        return
+
+    if data.switchState is None:
+        return
+    # update on change
+    updated_state = cast(SwitchState, data.switchState)
+    if updated_crownstone.state != updated_state.intensity:
+        updated_crownstone.state = updated_state.intensity
+
+        dispatcher_send(manager.hass, SIG_CROWNSTONE_STATE_UPDATE)
+
+
+def setup_sse_listeners(manager: CrownstoneEntryManager) -> None:
+    """Set up SSE listeners."""
+    # save unsub function for when entry removed
+    manager.listeners[SSE_LISTENERS] = [
+        manager.hass.bus.async_listen(
```

Comment:
We should not use the event bus for intra integration communication. Please use our dispatch helper for that instead.

Reply:
@RicArch97 Also keep in mind that events get logged to the database so if there are a lot of them they can decrease SD card lifetime for our RPi users.

Reply:
The eventbus is used with a specific use case in mind; The events can be used for device automation later, and specifically for the `event` trigger. Which creates a listener in the eventbus for the provided event type: https://github.com/home-assistant/core/blob/dev/homeassistant/components/homeassistant/triggers/event.py#L116

Which i have already implemented in our beta version; https://github.com/crownstone/crownstone-home-assistant/blob/master/custom_components/crownstone/device_trigger.py It is however only necessary for 1 event type, and not all event types. 

Of course this is irrelevant to this version, and i can easily replace it with dispatcher. I would however like to know if it's recommended to keep using the eventbus for that 1 specific event, with our beta version in mind.

Reply:
Device automations are a separate matter. For now we should use the dispatch helper if we need to signal another part of the integration.

We can still fire events for the user to use in automations.